### PR TITLE
Add Sparse fieldsets support

### LINKED ISSUE #1: RFC: JSON API Sparse Fieldsets
_From @bravo-kernel on April 13, 2017 18:47_

The JsonApiListener is currently missing [JSON API Sparse Fieldsets](http://jsonapi.org/format/#fetching-sparse-fieldsets) functionality.

> A client MAY request that an endpoint return only specific fields in the response on a per-type basis by including a fields[TYPE] parameter.

```
GET /articles?include=author&fields[articles]=title,body&fields[people]=name HTTP/1.1
Accept: application/vnd.api+json
```

_Copied from original issue: FriendsOfCake/crud#521_

### PR DESCRIPTION
**Pull request by rchavik**
To do:

- [x] test
- [x] pagination (once #29 is merged in)

Closes #1

### CODE REVIEW

**Comment by bravo-kernel** [comment]
http://jsonapi.org/examples/#sparse-fieldsets

**Comment by bravo-kernel** [comment]
Mr. @rchavik could pagination be required for the index page here as well?

**Comment by rchavik** [comment]
Yes, this has some dependency on #2. once that gets merged, i'll rebase this and adjust accordingly.

**Comment by rchavik** [comment]
This is still broken.

**Comment by bravo-kernel** [comment]
Do you mean the coverage (which seemed unrelated)?

**Comment by rchavik** [comment]
No, there's a bug somewhere.

I tried to fix it with cc02aec69e3d3522096006588b3617e707b52804 but it actually contained a typo and masked the problem.

File: src/Listener/JsonApiListener.php [hunk 1]
```diff
@@ -329,32 +329,35 @@ protected function _fieldSetsParameter($fieldSets, Subject $subject, $options)
         $fieldSets = array_map(function ($val) {
             return explode(',', $val);
         }, $fieldSets);
-        if (empty($fieldSets)) {
-            $this->config('fieldSets', $fieldSets);
-        }
```
Comment (rchavik) [inline src/Listener/JsonApiListener.php]: This was a typo, intended to be:

```php
if (!empty($fieldSets)) [
    $this->config('fieldSets', $fieldSets);
}
```

However, we call `$this->config('fieldSets', $fieldSets);`, it broke other things

Any insights on who/what/when to set this?

**Comment by rchavik** [comment]
This is now ready for review

File: src/Listener/JsonApiListener.php [hunk 2]
```diff
@@ -311,6 +311,59 @@ protected function _includeParameter($includes, Subject $subject, $options)
...
+                        return null;
+                    }
+
+                    return $repository->aliasField($val);
+                }, $fields);
+                $selectFields = array_merge($selectFields, array_filter($aliasFields));
+            }
+
+            $association = $associations->get($include);
+            if ($association) {
```
Comment (josegonzalez) [inline src/Listener/JsonApiListener.php]: !empty

**Review by josegonzalez — CHANGES_REQUESTED** [review]

File: src/Listener/JsonApiListener.php (hunk 2 above)
Comment (rchavik) [reply src/Listener/JsonApiListener.php]: I pushed the change you requested.

**Comment by bravo-kernel** [comment]
@josegonzalez are you ok with this PR now? If so I will merge.

**Review by bravo-kernel — APPROVED** [review]

**Review by josegonzalez — APPROVED** [review]
Needs docs probs, but good to go otherwise.

**Comment by bravo-kernel** [comment]
@josegonzalez yes, docs will appear somewhere this weekend as I think it will be quite a chore with testing and all.

Thanks for reviewing 💯

File: src/Listener/JsonApiListener.php [hunk 3]
```diff
@@ -311,6 +311,59 @@ protected function _includeParameter($includes, Subject $subject, $options)
...
+        $associations = $repository->associations();
+
+        $selectFields = [$repository->aliasField($repository->getPrimaryKey())];
+        $columns = $repository->schema()->columns();
+        $contains = [];
+        foreach ($fieldSets as $include => $fields) {
+            if ($include === Inflector::tableize($repository->alias())) {
+                $aliasFields = array_map(function ($val) use ($repository, $columns) {
+                    if (!in_array($val, $columns)) {
+                        return null;
```
Comment (dereuromark) [inline src/Listener/JsonApiListener.php]: This should be void return :) A sniffer should find this.

File: src/Listener/JsonApiListener.php [hunk 4]
```diff
@@ -311,6 +311,59 @@ protected function _includeParameter($includes, Subject $subject, $options)
...
+                    }
+
+                    return $repository->aliasField($val);
+                }, $fields);
+                $selectFields = array_merge($selectFields, array_filter($aliasFields));
```
Comment (dereuromark) [inline src/Listener/JsonApiListener.php:348]: If $aliasFields can be null, shouldnt you check on this before merging?

Comment (josegonzalez) [reply src/Listener/JsonApiListener.php:348]: Way to review after a merge :D

Comment (dereuromark) [reply src/Listener/JsonApiListener.php:348]: "merging" the variables, not the pull request :P

Comment (rchavik) [reply src/Listener/JsonApiListener.php:348]: Since `$aliasFields` is a result from an `array_map()`, it will always be an array, wouldn't it?

**Pull request merged**